### PR TITLE
fix moose interface to handle vector variables

### LINKED ISSUE #868: MOOSE Interface: Changing Single Value in String Vector
--------
Issue Description
--------
##### What did you expect to see happen?
My MOOSE input file has a function that changes a value at the beginning, middle, and end:
```
x = '0 1.0 2.0'
y = '1.8 1.95 2.0'
```
I would like to vary just the first number (1.8) using Raven.

##### What did you see instead?
The MOOSE interface currently has no way of allowing a single value (or multiple values) in a string to be identified as the value to modify.

##### Do you have a suggested fix for the development team?
Perhaps make an interface that takes multiple variables and joins them in a string separated by spaces. Some of the variables could be varying and others could be constant or dependent.  

I have, however, found some work arounds that are a bit complicated. It is possible to form the string using a variable with a `<function>` which returns a string. However, for an optimization problem, that is not possible currently. So for that I've been using an external model to form the string: `self.y_str = "'{0} 1.95 2.0'".format(self.varying[0])` But you have to go into `ExternalModel.py` and add `'str'` to the `self._availableVariableTypes` list. The index `[0]` is necessary as the external model has the varying variable in an array.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Update moose interface to handle vector variables, close #868 

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by wangcj05** [comment]
@mandd Do you have time to review this PR?

File: framework/CodeInterfaces/MooseBasedApp/MooseBasedAppInterface.py [hunk 1]
```diff
@@ -155,14 +154,21 @@ def _expandVarNames(self, **Kwargs):
     requests = []
     for var in Kwargs['SampledVars']:
       modifDict = {}
-      # When are colons used? I don't see any in the user manual example.
-      request, = var.split(':')
+      # colon is used when we want to perturb element in the vector of given variable
+      elemLoc = None
+      if ":" in var:
+        request, elemLoc = var.split(':')
```
Comment (mandd) [inline framework/CodeInterfaces/MooseBasedApp/MooseBasedAppInterface.py:160]: should we do also a strip() operation?

File: framework/JobHandler.py [hunk 2]
```diff
@@ -211,7 +211,7 @@ def __initializeRay(self):
             self.raiseADebug("Head host IP      :", address)
             self.raiseADebug("Head redis pass   :", redisPassword)
           ## Get servers and run ray remote listener
-          servers = self.runInfoDict['remoteNodes'] if self.rayInstanciatedOutside else self.__runRemoteListeningSockets(address, localHostName, redisPassword, db)
+          servers = self.runInfoDict['remoteNodes'] if self.rayInstanciatedOutside else self.__runRemoteListeningSockets(address, localHostName, redisPassword)
```
Comment (mandd) [inline framework/JobHandler.py:214]: is this modification part of this PR?

File: framework/CodeInterfaces/MooseBasedApp/MooseBasedAppInterface.py (hunk 1 above)
Comment (wangcj05) [reply framework/CodeInterfaces/MooseBasedApp/MooseBasedAppInterface.py]: updated.

File: framework/JobHandler.py (hunk 2 above)
Comment (wangcj05) [reply framework/JobHandler.py:214]: This is an error in our code, the 'db' is not a valid param for  `self.__runRemoteListeningSockets`

**Comment by wangcj05** [comment]
@mandd I have resolved your comments.

**Review by mandd — APPROVED** [review]

**Comment by mandd** [comment]
PR is good to me. Test board is green.

**Pull request merged**